File: app/src/main/java/org/senatov/cli/CliArgs.java
/*
 * CliArgs.java — CLI argument parser for MiMiComparator.
 * Supports: MiMiComparator <left> <right>
 *           MiMiComparator --left <path> --right <path>
 * Auto-detects file vs dir mode.
 * Iakov Senatov, 2026
 */
package org.senatov.cli;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;


@Slf4j
@Getter
public final class CliArgs {

    private final Path leftPath;
    private final Path rightPath;
    private final boolean autoCompare;
    private final boolean dirMode;


    private CliArgs(Path leftPath, Path rightPath) {
        this.leftPath = leftPath;
        this.rightPath = rightPath;
        this.autoCompare = leftPath != null && rightPath != null;
        this.dirMode = detectDirMode(leftPath, rightPath);
    }


    public static CliArgs parse(List<String> rawArgs) {
        log.info("parsing CLI args: {}", rawArgs);

        if (rawArgs == null || rawArgs.isEmpty()) {
            log.info("no CLI args — GUI chooser mode");
            return new CliArgs(null, null);
        }

        Path left = null;
        Path right = null;

        for (int i = 0; i < rawArgs.size(); i++) {
            String arg = rawArgs.get(i);

            if ("--left".equals(arg) && i + 1 < rawArgs.size()) {
                left = resolvePath(rawArgs.get(++i));
            } else if ("--right".equals(arg) && i + 1 < rawArgs.size()) {
                right = resolvePath(rawArgs.get(++i));
            } else if (!arg.startsWith("-") && left == null) {
                left = resolvePath(arg);
            } else if (!arg.startsWith("-") && right == null) {
                right = resolvePath(arg);
            } else {
                log.warn("ignoring unknown CLI arg: {}", arg);
            }
        }

        log.info("parsed CLI: left={} right={} autoCompare={}",
                left, right, left != null && right != null);
        return new CliArgs(left, right);
    }


    public Optional<Path> left() {
        return Optional.ofNullable(leftPath);
    }


    public Optional<Path> right() {
        return Optional.ofNullable(rightPath);
    }


    private static Path resolvePath(String raw) {
        Path path = Path.of(raw).toAbsolutePath().normalize();

        if (!Files.exists(path)) {
            log.warn("CLI path does not exist: {}", path);
            return null;
        }

        return path;
    }


    private static boolean detectDirMode(Path left, Path right) {
        if (left != null && Files.isDirectory(left)) {
            return true;
        }

        if (right != null && Files.isDirectory(right)) {
            return true;
        }

        return false;
    }
}

File: app/src/main/kotlin/org/senatov/cli/CliArgs.kt
/*
 * CliArgs — CLI argument parser 4 MiMiComparator.
 * MiMiComparator <left> <right>
 * MiMiComparator --left <path> --right <path>®
 * auto-detects file vs dir mode.
 * Iakov Senatov, 2026
 */
package org.senatov.cli

import org.senatov.helpers.log.LogHelper
import org.senatov.helpers.log.LogTag
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Path


class CliArgs private constructor(
    val leftPath: Path?,
    val rightPath: Path?,
    val autoCompare: Boolean,
    val isDirMode: Boolean,
    private val dirModeExplicit: Boolean
) {

    fun left(): Path? {
        LogHelper.enter(log, LogTag.CLI, "left")
        return leftPath
    }

    fun right(): Path? {
        LogHelper.enter(log, LogTag.CLI, "right")
        return rightPath
    }

    fun hasExplicitDirMode(): Boolean {
        LogHelper.enter(log, LogTag.CLI, "hasExplicitDirMode")
        return dirModeExplicit
    }


    companion object {
        private val log = LoggerFactory.getLogger(CliArgs::class.java)

        fun parse(rawArgs: List<String>?): CliArgs {
            LogHelper.enter(log, LogTag.CLI, "parse", "rawArgs" to rawArgs)
            log.debug(LogTag.CLI, "parse args={}", rawArgs?.size ?: 0)
            if (rawArgs.isNullOrEmpty()) {
                log.info(LogTag.CLI, "GUI mode")
                return CliArgs(null, null, false, false, false)
            }
            var left: Path? = null
            var right: Path? = null
            var forcedDirMode: Boolean? = null
            var i = 0
            while (i < rawArgs.size) {
                val arg = rawArgs[i]
                when {
                    arg == "--left" && i + 1 < rawArgs.size -> { left = resolvePath(rawArgs[++i]) }
                    arg == "--right" && i + 1 < rawArgs.size -> { right = resolvePath(rawArgs[++i]) }
                    arg == "--dirs" -> forcedDirMode = true
                    arg == "--files" -> forcedDirMode = false
                    !arg.startsWith("-") && left == null -> left = resolvePath(arg)
                    !arg.startsWith("-") && right == null -> right = resolvePath(arg)
                    else -> log.warn(LogTag.CLI, "ignored arg {}", arg)
                }
                i++
            }
            val dirMode = forcedDirMode ?: detectDirMode(left, right)
            val auto = left != null && right != null
            log.info(LogTag.CLI, "parsed left={} right={} auto={} dir={} explicit={}",
                left, right, auto, dirMode, forcedDirMode != null)
            return CliArgs(left, right, auto, dirMode, forcedDirMode != null)
        }


        private fun resolvePath(raw: String): Path? {
            LogHelper.enter(log, LogTag.CLI, "resolvePath", "raw" to raw)
            val path = Path.of(raw).toAbsolutePath().normalize()
            if (!Files.exists(path)) {
                log.warn(LogTag.CLI, "path missing {}", path)
                return null
            }
            return path
        }

        private fun detectDirMode(left: Path?, right: Path?): Boolean {
            LogHelper.enter(log, LogTag.CLI, "detectDirMode", "left" to left, "right" to right)
            if (left != null && Files.isDirectory(left)) return true
            if (right != null && Files.isDirectory(right)) return true
            return false
        }
    }
}